runtime/src/main/java/org/corfudb/runtime/Restore.java

```diff
@@ -0,0 +1,165 @@
+            } catch (IOException e) {
+                log.error("failed to restore table {} from temp file {}", streamId, fileName);
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Restore a single table
+     *
+     * @param fileName   - the name of the temp backup file
+     * @param streamId   - the stream ID of the table which is to be restored
+     * @throws IOException
+     */
+    private void restoreTable(String fileName, UUID streamId) throws IOException {
+        try (FileInputStream fileInput = new FileInputStream(fileName)) {
+            // Clear table before restore
+            TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE);
+            SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+            txn.logUpdate(streamId, entry);
+
+            long numEntries = 0;
+            while (fileInput.available()> 0) {
+                OpaqueEntry opaqueEntry = OpaqueEntry.read(fileInput);
+                List<SMREntry> smrEntries = opaqueEntry.getEntries().get(streamId);
+                if (smrEntries == null || smrEntries.isEmpty()) {
+                    continue;
+                }
+
+                txn.logUpdate(streamId, smrEntries);
+                numEntries++;
+            }
+            txn.commit();
+            log.debug("write uuid {} with {} numEntries", streamId, numEntries);
+        } catch (FileNotFoundException e) {
+            log.error("restoreTable can not find file {}", fileName);
+            throw e;
+        }
+    }
+
+    /**
+     * Open the backup tar file and save the table backups to tableDir directory
+     */
+    private void openTarFile() throws IOException {
+        try (FileInputStream fileInput = new FileInputStream(filePath);
+             TarArchiveInputStream tarInput = new TarArchiveInputStream(fileInput)) {
+            getTablesFromTarFile(tarInput);
+        } catch (IOException e) {
+            log.error("failed to get tables from tar file {}", filePath);
+            throw e;
+        }
+    }
+
+    private void getTablesFromTarFile(TarArchiveInputStream tarInput) throws IOException {
+        int count;
+        byte[] buf = new byte[1024];
+        TarArchiveEntry entry;
+        while ((entry = tarInput.getNextTarEntry()) != null) {
+            streamIDs.add(UUID.fromString(entry.getName()));
+            String tablePath = restoreTempDirPath + File.separator + entry.getName();
+            try (FileOutputStream fos = new FileOutputStream(tablePath)) {
+                while ((count = tarInput.read(buf, 0, 1024)) != -1) {
+                    fos.write(buf, 0, count);
+                }
+            }
+        }
+    }
+
+    /**
+     * Some verification logic (TBD) such as
+     * - Compare the user provided streamIds and names of table backups under tmp directory, or some metadata file
+     * - Checksum
+     */
+    private void verify() {
+        // TODO: verification logic to be implemented
```

Comment:
I think this actually has to do with my comment.. should we solve this now?

Reply:
I haven't decided what to check as verification. Also since it's a nice-to-have feature and not currently asked for by Graham's team, I think it might be better to not put too many efforts in this very first backup restore PR.